---
title: "Class 11: Structural Bioinformatics pt2"
author: "Le, Sarah (A18518276)"
format: html
toc: true
---

## AlphaFold Data Base (AFDB)

The EBI maintains the largest database of AlphaFold structure prediction models at: https://alphafold.ebi.ac.uk/

From last class (before Halloween), we saw that the PDB had 244,290 (Oct 2025).

The total number of protein sequences in UniProtKB is 199,579,901. 

> **Key Point**: This is a tiny fraction of sequence space that has structural coverage. (0.12%)

```{r}
244290/ 199579901* 100
```

AFDB is attempting to address this gap...

There are two "Quality Scores" from AlphaFold. One for residues (i.e. each amino acid) called **pLDDT** score. The other **Predicted Aligned Error (PAE)** measures the confidence in the relative position of two residues (i.e. a score for every pair of residues).

## Generating your own structure predictions

Figure of 5 generated HIV-PR models

![](HIVPR_DIMER_23119_UNRELAXED_RANK_001_ALPHAFOLD2_MULTIMER_V3_MODEL_4_SEED_000.PDB-HIVPR_DIMER_23119_UNRELAXED_RANK_002_ALPHAFOLD2_MULTIMER_V3_MODEL_1_SEED_000.PDB-HIVPR_DIMER_23119_UNRELAXED_RANK_003_ALPHAFOLD2_MULTIMER_V3_MODEL_5_SEED_000.P full.png) 

And the top ranked model colored by chain.

![](HIVPR_DIMER_topmodel.png)

pLDDT score for model 1.

![](HIVPR_DIMER_PLDDT_m1.png)

and pLDDT score for model 5.

![](HIVPR_DIMER_PLDDT_m5.png)

## Custom analysis of resulting models in R

Read key result file into R. The first thing I need to know is what my results directory/ folder is called (i.e. its name is different for every AlphaFold run/ job)

```{r}
results_dir <- "HIVPR_dimer_23119/"

# File names for all PDB models
pdb_files <- list.files(path=results_dir,
                        pattern="*.pdb",
                        full.names = TRUE)

# Print our PDB file names
basename(pdb_files)
```

```{r}
library(bio3d)

m1 <- read.pdb(pdb_files[1])
m1
```

```{r}
head(m1$atom)
```

```{r}
m1$atom$b[m1$calpha]
```

```{r}
plot(m1$atom$b[m1$calpha], typ= "l", ylim= c(0,100))
```

```{r}
plot.bio3d(m1$atom$b[m1$calpha], typ= "l")
```

## Residue conservation from alignment file

Find the large AlphaFold alignment file.

```{r}
aln_file <- list.files(path=results_dir,
                       pattern=".a3m$",
                        full.names = TRUE)
aln_file
```

Read this into R.

```{r}
aln <- read.fasta(aln_file[1], to.upper = TRUE)
```

How many sequences are in this alignment

```{r}
dim(aln$ali)
```

We can score residue conservation in the alignment with the `conserv()` function.

```{r}
sim <- conserv(aln)
```

```{r}
plotb3(sim[1:99], sse=trim.pdb(m1, chain="A"),
       ylab="Conservation Score")
```

```{r}
con <- consensus(aln, cutoff = 0.9)
con$seq
```

For a final visualization of these functionally important sites we can map this conservation score to the Occupancy column of a PDB file for viewing in molecular viewer programs such as Mol*, PyMol, VMD, chimera etc.

```{r}
m1.pdb <- read.pdb(pdb_files[1])
occ <- vec2resno(c(sim[1:99], sim[1:99]), m1.pdb$atom$resno)
write.pdb(m1.pdb, o=occ, file="m1_conserv.pdb")
```

Data generated from and Mol* using coloring by Occupancy with top ranked dimer model colored by sequence conservation. Conserved positions in a darker purple. The DTGA motif of one chain is highlighted in green.

![](M1_CONSERV.PDB.png)